Firestore/core/src/firebase/firestore/util/shared_value.h

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_SHARED_VALUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_SHARED_VALUE_H_
+
+#include <memory>
+#include <type_traits>
+
+#include "absl/meta/type_traits.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * A smart pointer that implements shared ownership but otherwise with value
+ * semantics. That is, operator== compares the thing pointed to, not the pointer
+ * itself.
+ *
+ * Another way to look at this is that it's like std::optional but the
+ * underlying value is shared instead of copied.
+ */
+template <typename T>
+class shared_value {
```

Comment:
Most important question: is this thing even worthwhile?

The alternative would be to have retain custom equality checkers in Query.

I can't help but shake the feeling that this is semantically sort of muddy and doesn't really conceptually make sense.

Reply:
I don't have a very strong opinion on this. I'm not against adding this class, but it does seem like it's overkill for implementing a three-line function. Also, it doesn't necessarily prevent comparison errors because the code author still has to remember to use `shared_value` instead of `shared_ptr` (unless we want to completely ban `shared_ptr`, which seems excessive).

As an alternative: how feasible is it to write a `clang-tidy` check that flags pointers/smart pointers being checked for equality?